src/Microsoft.OData.Core/Uri/NodeToStringBuilder.cs

```diff
@@ -948,5 +730,95 @@ private static bool IsValidSearchWord(string text)
                 return true;
             }
         }
+
+        /// <summary>
+        /// Check whether Navigation source of the FilterClause rangeVariable is different from the Expression rangeVariable.
+        /// </summary>
+        /// <param name="node">Expression node.</param>
+        /// <returns>If Navigation Source are different, returns true. Otherwise false.</returns>
+        private bool IsDifferentSource(QueryNode node)
+        {
+            ResourceRangeVariableReferenceNode rangeVariableReferenceNode = GetResourceRangeVariableReferenceNode(node);
+            if (rangeVariableReferenceNode == null)
+            {
+                return false;
+            }
+            if (rangeVariables.Count == 0)
+            {
+                return false;
+            }
+
+            RangeVariable rangeVariable = rangeVariables.Peek();
+            if (rangeVariable == null)
+            {
+                return false;
+            }
+            return rangeVariable is ResourceRangeVariable resourceRangeVariable ? resourceRangeVariable.NavigationSource != rangeVariableReferenceNode.NavigationSource : false;

```

Comment:
Since we will be adding support for $root and $this range variables, check if the `rangeVariableReferenceNode.Name == ExpressionConstants.It`

`return rangeVariable is ResourceRangeVariable resourceRangeVariable ? resourceRangeVariable.NavigationSource != rangeVariableReferenceNode.NavigationSource && rangeVariableReferenceNode.Name == ExpressionConstants.It : false;`